spatialprofilingtoolbox/db/study_access.py

```diff
@@ -0,0 +1,228 @@
+"""Convenience accessors of study-related small data / metadata."""
+from typing import cast
+import re
+
+from spatialprofilingtoolbox.workflow.common.export_features import ADIFeatureSpecificationUploader
+from spatialprofilingtoolbox.db.exchange_data_formats.study import StudyContact
+from spatialprofilingtoolbox.db.exchange_data_formats.study import DataRelease
+from spatialprofilingtoolbox.db.exchange_data_formats.study import Publication
+from spatialprofilingtoolbox.db.exchange_data_formats.study import Institution
+from spatialprofilingtoolbox.db.exchange_data_formats.study import Assay
+from spatialprofilingtoolbox.db.exchange_data_formats.study import CountsSummary
+from spatialprofilingtoolbox.db.exchange_data_formats.study import StudyComponents
+from spatialprofilingtoolbox.db.exchange_data_formats.study import StudyHandle
+from spatialprofilingtoolbox.db.exchange_data_formats.study import StudySummary
+from spatialprofilingtoolbox.db.exchange_data_formats.study import Context
+from spatialprofilingtoolbox.db.exchange_data_formats.study import Products
+from spatialprofilingtoolbox.db.simple_query_patterns import GetSingleResult
+from spatialprofilingtoolbox.db.cohorts import _get_sample_cohorts
+from spatialprofilingtoolbox.db.database_connection import SimpleReadOnlyProvider
+
+class StudyAccess(SimpleReadOnlyProvider):
```

Comment:
Why is this class separate from the `Query` class? They're both initialized with the same parameter (study) and return simple queries on the study.

Reply:
To directly answer your question: `StudyAccess` is initialized with `cursor`, not `study`, and `QueryHandler` (which is what I think you meant by `Query`) is a delegator not just to `StudyAccess` but to several other classes.

What's going on here is the following...
There are classes `StudyAccess`, `PhenotypeAccess`, `FractionsAccess`, and `UMAPAccess` which are responsible for querying the database and giving back data with a controlled type, each doing their work in their own scope of responsibility. Each one subclasses `SimpleReadOnlyProvider`, for the purpose of having a `cursor` member readily available to each method without requiring laborious repetition of `cursor` as the first argument to every method.
In order to relieve the FastAPI application routes from having to deal with cursors, and with delegation to different scopes of responsibility, there are middlemen `QueryHandler`(stateless, class methods only), and `QueryCursor` (handles context manager for cursor), and the latter is finally exposed. This is how we get nice syntax like:
```py
@app.get("/phenotype-criteria/")
async def get_phenotype_criteria(
    study: ValidStudy,
    phenotype_symbol: ValidPhenotypeSymbol,
) -> PhenotypeCriteria:
    return query().get_phenotype_criteria(study, phenotype_symbol)
```
which would otherwise look like:
```py
@app.get("/phenotype-criteria/")
async def get_phenotype_criteria(
    study: ValidStudy,
    phenotype_symbol: ValidPhenotypeSymbol,
) -> PhenotypeCriteria:
    with DBCursor as cursor():
        PhenotypesAccess(cursor).get_phenotype_criteria(study, phenotype_symbol)
```
To make this work I have `QueryCursor` add methods to itself on-the-fly during initialization, so mypy was unhappy. To make mypy happy I added explicit redundant declarations of the method names:
```py
get_study_components: Callable
retrieve_study_handles: Callable
get_number_cells: Callable
get_study_summary: Callable
get_cell_fractions_summary: Callable
get_phenotype_symbols: Callable
get_phenotype_symbols_all_studies: Callable
get_composite_phenotype_identifiers: Callable
get_phenotype_criteria: Callable
get_channel_names_all_studies: Callable
retrieve_signature_of_phenotype: Callable
get_umaps_low_resolution: Callable
get_umap: Callable
```
I'm open to alternative patterns for achieving the same simplicity of caller syntax.